File: src/ruby/comb_sort.rb
# frozen_string_literal: true

# Algoritmo de Comb sort em Ruby
# Carlos Alves
# https://github.com/EuCarlos

def comb_sort(arr)
  gap = arr.length
  shrink = 1.3
  swapped = true

  while gap > 1 || swapped
    gap = (gap / shrink).floor
    gap = 1 if gap < 1
    swapped = false

    (0..(arr.length - gap) - 1).each do |i|
      if arr[i] > arr[i + gap]
        arr[i], arr[i + gap] = arr[i + gap], arr[i]
        swapped = true
      end
    end
  end

  arr
end

array = [86, 16, 10, 39, 49, 6, 32, 57, 72, 29]
print "Lista: #{array} \n"
print "Lista Ordenada: #{comb_sort(array)}"

File: src/go/comb_sort.go
package main

import "fmt"

func CombSort(slice []int) {
	shrinkFactor := 1.247330950103979
	gap := len(slice)
	swapped := true
	i := 0

	for gap > 1 || swapped {
		if gap > 1 {
			gap = int(float64(gap) / shrinkFactor)
		}

		swapped = false
		i = 0

		for gap+i < len(slice) {
			if slice[i]-slice[i+gap] > 0 {
				slice[i], slice[i+gap] = slice[i+gap], slice[i]
				swapped = true
			}
			i++
		}
	}
}

func main() {
	slice := []int{6, 5, 8, 7, 3, 4, 2, 1, 9}
	fmt.Println("Slice:", slice)
	CombSort(slice)
	fmt.Println("CombSort:", slice)
}
